src/ctest.ts

```diff
@@ -602,14 +604,40 @@ export class CTestDriver implements vscode.Disposable {
         return uniqueTests;
     }
 
-    private async runTestHandler(request: vscode.TestRunRequest, cancellation: vscode.CancellationToken, useCMakePresets: boolean) {
+    // Returns false if test preset wasn't selected already (changing test preset triggers test explorer refresh)
+    private async checkTestPreset(tests: vscode.TestItem[]): Promise<boolean> {
+        let presetMayChange = false;
+        for (const test of tests) {
+            if (test.id === testPresetRequired) {
+                const folder = test.parent ? test.parent.id : test.id;
+                const project = await this.projectController?.getProjectForFolder(folder);
+                if (!project) {
+                    log.error(localize('no.project.found', 'No project found for folder {0}', folder));
+                    return false;
+                }
+                await vscode.commands.executeCommand('cmake.selectTestPreset', project.workspaceFolder);
```

Comment:
Doesn't this prompt for a selection?  I thought this method just wants to "check" presets, not set them.  It might be cleaner to check the result of calling this method and then prompt to select the test preset if it fails.

Reply:
See my reply to the comment below, this function relies on the test id to get the folder name and other info. Separating it is rather complicated. I'll try to find a proper name for it and add some comments.

Reply:
Does `test` represent a folder, or an individual test?  If it's a workspace folder, you might want to rename it.  If it's a single test, then the current implementation will ask you to set the preset for every single test in the folder which I don't think makes sense.

Reply:
When a test preset is not selected, as shown in the screenshot above, there will only be one "test" under each folder. This "test" has a special ID. If the test ID matches this special ID, user will be prompted to select a test preset for this folder.

Reply:
I'm confused. Is it a test, or a node in the test explorer with tests under it?

Reply:
When no presets are selected, we can't know the correct test list. So the structure will be
```
- Folder 1
  - Select a test preset to discover tests <- this is the special "test"
- Folder 2
  - ...
```

Reply:
So `test` is the placeholder for the folder?  Can you say that in a comment when we enter this `if` block?

Reply:
It's a placeholder entry for the list of test in the folder. But I think you know what I mean. Sure I'll add a comment.